api/src/main/kotlin/com/google/devtools/ksp/processing/CodeGenerator.kt

```diff
@@ -32,22 +32,38 @@ interface CodeGenerator {
     /**
      * Creates a file which is managed by [CodeGenerator]
      *
+     * [sources] of corresponding [KSNode]s which are obtained directly from [Resolver] need to be specified.
+     * Namely, the containing files of those [KSNode]s who are obtained from:
+     *   * [Resolver.getAllFiles]
+     *   * [Resolver.getSymbolsWithAnnotation]
+     *   * [Resolver.getClassDeclarationByName]
+     *
+     * Instead of requiring processors to specify all source files which are relevant in generating the given output,
+     * KSP traces dependencies automatically and only need to know those sources that only processors know what they
+     * are for. If a [KSFile] is indirectly obtained through other [KSNode]s, it hasn't to be specified for the given
+     * output, even if its contents contribute to the generation of the output.
+     *
+     * For example, a processor generates an output `O` after reading class `A` in `A.kt` and class `B` in `B.kt`,
+     * where `A` extends `B`. The processor got `A` by [Resolver.getSymbolsWithAnnotation] and then got `B` by
+     * [KSClassDeclaration.superTypes] from `A`. Because the inclusion of `B` is due to `A`, `B.kt` needn't to be
+     * specified in [sources] for `O`. Note that specifying `B.kt` in this case doesn't hurt, it is only unnecessary.
+     *
+     * @param sources are [KSFile]s from which this output is built. Only those that are obtained directly from [Resolver]
+     *                are required.
      * @param packageName corresponds to the relative path of the generated file; using either '.'or '/' as separator.
      * @param fileName file name
      * @param extensionName If "kt" or "java", this file will participate in subsequent compilation.
      *                      Otherwise its creation is only considered in incremental processing.
      * @return OutputStream for writing into files.
      * @see [CodeGenerator] for more details.
      */
-    fun createNewFile(packageName: String, fileName: String, extensionName: String = "kt"): OutputStream
+    fun createNewFile(sources: List<KSFile>, packageName: String, fileName: String, extensionName: String = "kt"): OutputStream
```

Comment:
This is a breaking change that is only needed when incremental is enabled, can you make it default to empty list?

Reply:
actually this is a bit scary to set to empty list because that would mean the processor will be silently broken in incremental compilation.
Also, we cannot distinguish whether developer just doesn't have additional sources OR they didn't implement incremental compilation.

Should we instead provide an OPT-OUT API where processor needs to specify that it doesn't support incremental compilation and we throw an error if incremental processing is enabled? (or a warning if we can workaround it by invalidating it every time)
That being said, i'm all for making opt out harder :D.

Reply:
Yeah, this is intended to make opt-out hard :p How about a AllFilesWildcard with a scary comment "this will disable incremental processing"?

Reply:
yea i like that :)
does that mean this should change to a specific class like `Dependencies` which can be
`Dependencies(files)` or `Dependencies.ALL_FILES`
Not sure if dependencies is a good name but making it another class would give developer the attention (otherwise, it is easy to pass empty list and not think about it)

Reply:
@yigit @neetopia Updated. Please take a look again.